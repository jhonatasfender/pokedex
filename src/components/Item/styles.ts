import styled from "styled-components";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import Image from "../Image";

export const Img = styled(Image)`
  position: absolute;
  width: 15rem;
  height: 15rem;
  top: -1rem;
  right: 0.5rem;
  z-index: 1;
`;

type LikeProps = {
  fixed: boolean;
};

export const Like = styled(FontAwesomeIcon)<LikeProps>`
  position: absolute;
  cursor: pointer;
  width: ${({ fixed }) => (fixed ? "2rem" : "0")};
  height: ${({ fixed }) => (fixed ? "2rem" : "0")};
  bottom: 1rem;
  right: 1rem;
  z-index: 3;
  transition: all 0.2s ease-in-out;
  transition-delay: 0.2s;
`;

export const WrapperItem = styled.div`
  --color: ${({ theme }) => theme.colors.primary};

  display: inherit;
  flex-direction: column;
  position: relative;
  gap: 1rem;
  width: 26rem;
  height: 10rem;
  background-color: var(--color);
  border-radius: 0.8rem;
  padding: 1rem;
  box-shadow: 0 0 1.3rem -0.3rem var(--color);
  overflow: hidden;
  user-select: none;

  &:hover {
    ${Like} {
      width: 2rem;
      height: 2rem;
    }
  }
`;

export const NationalNumber = styled.span`
  position: absolute;
  top: 1rem;
  right: 1rem;
  z-index: 2;

  font-size: 1.4rem;
  font-weight: bold;
  opacity: 0.3;
`;

export const Title = styled.span`
  color: ${({ theme }) => theme.colors.secondary};
  font-size: 1.8rem;
  font-weight: bold;
`;

export const Type = styled.div`
  color: ${({ theme }) => theme.colors.secondary};
  background-color: ${({ theme }) => theme.colors.turquoise};
  border-radius: 1rem;
  padding: 0.5rem;
  width: 5rem;
  text-align: center;
  font-weight: bold;
`;
