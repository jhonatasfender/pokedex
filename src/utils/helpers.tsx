import React from "react";
import { ThemeProvider } from "styled-components";

import { render, RenderResult } from "@testing-library/react";

import theme from "../style-global/theme";

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
