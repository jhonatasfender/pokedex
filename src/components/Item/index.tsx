import { faHeart } from "@fortawesome/free-solid-svg-icons";

import { faHeart as farHeart } from "@fortawesome/free-regular-svg-icons";

import { Pokemon } from "../../hooks/use-pokemon";
import * as S from "./styles";

type ItemProps = {
  item: Pokemon;
  onLike: (item: Pokemon) => void;
};

export default function Item({ item, onLike }: ItemProps) {
  return (
    <S.WrapperItem>
      <S.NationalNumber>{item.national_number}</S.NationalNumber>
      <S.Title>{item.name}</S.Title>

      <S.Img sprites={item.sprites} />

      {item.type.map((type) => (
        <S.Type key={type}>{type}</S.Type>
      ))}

      <S.Like
        aria-hidden={false}
        fixed={!!item.like}
        icon={item.like ? faHeart : farHeart}
        onClick={() => onLike(item)}
      />
    </S.WrapperItem>
  );
}
