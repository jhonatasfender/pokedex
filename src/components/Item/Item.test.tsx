import { logRoles, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { renderWithTheme } from "../../utils/helpers";
import Item from "./";

const inDefault = {
  national_number: "001",
  evolution: null,
  sprites: {
    normal:
      "https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.png",
    large: "https://img.pokemondb.net/artwork/bulbasaur.jpg",
    animated:
      "https://img.pokemondb.net/sprites/black-white/anim/normal/bulbasaur.gif",
  },
  name: "Bulbasaur",
  type: ["Grass", "Poison"],
  total: 318,
  hp: 45,
  attack: 49,
  defense: 49,
  sp_atk: 65,
  sp_def: 65,
  speed: 45,
};

describe("Header", () => {
  it("should initialize with default parameters", () => {
    const onLike = jest.fn();

    renderWithTheme(<Item item={inDefault} onLike={onLike} />);

    const card = screen.getByRole("card");

    expect(card).toMatchSnapshot();
  });

  it("should initialize with default parameters with liked", () => {
    const onLike = jest.fn();

    renderWithTheme(
      <Item item={{ ...inDefault, like: true }} onLike={onLike} />
    );

    const card = screen.getByRole("card");

    expect(card).toMatchSnapshot();
  });

  it("should call a like function when clicking the button", async () => {
    const onLike = jest.fn();

    renderWithTheme(
      <Item item={{ ...inDefault, like: true }} onLike={onLike} />
    );

    const card = screen.getByRole("card");
    userEvent.hover(card);

    const [, img] = screen.getAllByRole("img");
    userEvent.click(img);

    await waitFor(() => {
      expect(onLike).toHaveBeenCalledWith({ ...inDefault, like: true });
    });
  });
});
