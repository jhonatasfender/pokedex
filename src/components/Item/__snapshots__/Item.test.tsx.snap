// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should initialize with default parameters 1`] = `
.c3 {
  position: absolute;
  width: 15rem;
  height: 15rem;
  top: -1rem;
  right: 0.5rem;
  z-index: 1;
}

.c6 {
  position: absolute;
  cursor: pointer;
  width: 0;
  height: 0;
  bottom: 1rem;
  right: 1rem;
  z-index: 3;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  -webkit-transition-delay: 0.2s;
  transition-delay: 0.2s;
}

.c0 {
  --color: #3ed3af;
  display: inherit;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  gap: 1rem;
  width: 26rem;
  height: 10rem;
  background-color: var(--color);
  border-radius: 0.8rem;
  padding: 1rem;
  box-shadow: 0 0 1.3rem -0.3rem var(--color);
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0:hover .c5 {
  width: 2rem;
  height: 2rem;
}

.c1 {
  position: absolute;
  top: 1rem;
  right: 1rem;
  z-index: 2;
  font-size: 1.4rem;
  font-weight: bold;
  opacity: 0.3;
}

.c2 {
  color: #FFFFFF;
  font-size: 1.8rem;
  font-weight: bold;
}

.c4 {
  color: #FFFFFF;
  background-color: #27f1c0;
  border-radius: 1rem;
  padding: 0.5rem;
  width: 5rem;
  text-align: center;
  font-weight: bold;
}

<div
  class="c0"
  role="card"
>
  <span
    class="c1"
  >
    001
  </span>
  <span
    class="c2"
  >
    Bulbasaur
  </span>
  <img
    alt="pokemon"
    class="c3"
    src="https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.png"
  />
  <div
    class="c4"
  >
    Grass
  </div>
  <div
    class="c4"
  >
    Poison
  </div>
  <svg
    aria-hidden="false"
    class="svg-inline--fa fa-heart c5 c6"
    data-icon="heart"
    data-prefix="far"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M244 84L255.1 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 0 232.4 0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84C243.1 84 244 84.01 244 84L244 84zM255.1 163.9L210.1 117.1C188.4 96.28 157.6 86.4 127.3 91.44C81.55 99.07 48 138.7 48 185.1V190.9C48 219.1 59.71 246.1 80.34 265.3L256 429.3L431.7 265.3C452.3 246.1 464 219.1 464 190.9V185.1C464 138.7 430.4 99.07 384.7 91.44C354.4 86.4 323.6 96.28 301.9 117.1L255.1 163.9z"
      fill="currentColor"
    />
  </svg>
</div>
`;

exports[`Header should initialize with default parameters with liked 1`] = `
.c3 {
  position: absolute;
  width: 15rem;
  height: 15rem;
  top: -1rem;
  right: 0.5rem;
  z-index: 1;
}

.c6 {
  position: absolute;
  cursor: pointer;
  width: 2rem;
  height: 2rem;
  bottom: 1rem;
  right: 1rem;
  z-index: 3;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  -webkit-transition-delay: 0.2s;
  transition-delay: 0.2s;
}

.c0 {
  --color: #3ed3af;
  display: inherit;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  gap: 1rem;
  width: 26rem;
  height: 10rem;
  background-color: var(--color);
  border-radius: 0.8rem;
  padding: 1rem;
  box-shadow: 0 0 1.3rem -0.3rem var(--color);
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0:hover .c5 {
  width: 2rem;
  height: 2rem;
}

.c1 {
  position: absolute;
  top: 1rem;
  right: 1rem;
  z-index: 2;
  font-size: 1.4rem;
  font-weight: bold;
  opacity: 0.3;
}

.c2 {
  color: #FFFFFF;
  font-size: 1.8rem;
  font-weight: bold;
}

.c4 {
  color: #FFFFFF;
  background-color: #27f1c0;
  border-radius: 1rem;
  padding: 0.5rem;
  width: 5rem;
  text-align: center;
  font-weight: bold;
}

<div
  class="c0"
  role="card"
>
  <span
    class="c1"
  >
    001
  </span>
  <span
    class="c2"
  >
    Bulbasaur
  </span>
  <img
    alt="pokemon"
    class="c3"
    src="https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.png"
  />
  <div
    class="c4"
  >
    Grass
  </div>
  <div
    class="c4"
  >
    Poison
  </div>
  <svg
    aria-hidden="false"
    class="svg-inline--fa fa-heart c5 c6"
    data-icon="heart"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
      fill="currentColor"
    />
  </svg>
</div>
`;
